### PR TITLE
POC for Kerberos on Windows with SSPI, take 2

### PR DESCRIPTION
**Pull request by vkarpov15**
Unfortunately can't change branches on a Github PR AFAIK, for the sake of sanity this is #20 against the v2-dev branch

### CODE REVIEW

**Comment by niemeyer** [comment]
There are comments in #20 that were not addressed. Even some of the ones marked as "Done".

**Comment by vkarpov15** [comment]
Yeah I hadn't pushed them yet, sorry. Subsequent commit contains them. I just wanted to open up a new PR because I merged v2-dev into this branch and so the previous PR had a lot of noise.

**Comment by niemeyer** [comment]
Okay, please let me know once this is actually ready for review.

**Comment by vkarpov15** [comment]
Go ahead and review, I'm reasonably certain I've addressed all your comments.

File: sasl/sasl.go [hunk 1]
```diff
@@ -2,6 +2,9 @@
 //
 // This package is not meant to be used by itself.
 //
+
+// +build linux darwin
```
Comment (niemeyer) [inline sasl/sasl.go]: // +build !windows

File: sasl/sasl_windows.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	return cstr
+}
+
+func (ss *saslSession) Close() {
+	for _, cstr := range ss.stringsToFree {
+		C.free(unsafe.Pointer(cstr))
+	}
+
+	// Make sure we've cleaned up all the buffers we malloced when we're sure we don't need em anymore
+	if ss.authComplete || ss.errored {
```
Comment (niemeyer) [inline sasl/sasl_windows.go]: This was discussed in the prior review, and still looks pending.

File: sasl/sasl_windows.c [hunk 3]
```diff
@@ -0,0 +1,126 @@
+#ifdef _WIN32
```
Comment (niemeyer) [inline sasl/sasl_windows.c]: I don't think this file will get built when not on windows, given its name.

File: sasl/sasl_windows.h [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#ifdef _WIN32
```
Comment (niemeyer) [inline sasl/sasl_windows.h]: I don't think this file will get built when not on windows, given its name.

**Comment by niemeyer** [comment]
Ok, just did a quick scan with a few comments, but I'll need to go back to it later.

Meanwhile, where are the tests?

**Comment by vkarpov15** [comment]
Thanks for comments, I'll fix them in a bit. Re: tests, working on that, that will require moving kerb tests to a separate suite for now. My current testing setup overwrites existing tests and so isn't mergable.

File: sasl/sasl_windows.h (hunk 4 above)
Comment (vkarpov15) [inline sasl/sasl_windows.h]: done

File: sasl/sasl_windows.c (hunk 3 above)
Comment (vkarpov15) [inline sasl/sasl_windows.c]: done

File: sasl/sasl_windows.go (hunk 2 above)
Comment (vkarpov15) [inline sasl/sasl_windows.go]: done

File: sasl/sasl.go (hunk 1 above)
Comment (vkarpov15) [inline sasl/sasl.go]: done

**Comment by vkarpov15** [comment]
Hey Gustavo, just a gentle reminder to see if you have any more comments.

**Comment by niemeyer** [comment]
I might, but where are the tests?

**Comment by vkarpov15** [comment]
In a separate branch and ready to merge into this one. Do you want them in
this branch? There's some significant work there to break Kerberos tests
off into their own suite so they don't go through the normal setup and
teardown.

On Mon, Sep 22, 2014 at 12:30 PM, Gustavo Niemeyer <notifications@github.com

**Comment by niemeyer** [comment]
I don't understand what kind of significant work might take place for that. Registering a new test suite requires three or four lines of code.

Either way, it doesn't make sense to merge this without tests. Please also make sure that the tests that are being submitted with the logic actually exercise it well, and are passing for real.

**Comment by niemeyer** [comment]
Also, what do you mean by "break Kerberos tests"?  The test that exists today are trivial in their implementation and design. We shouldn't be breaking them out or removing them from where they are.

**Comment by niemeyer** [comment]
It might be best to submit that new logic you're talking about in a separate branch before polluting this PR, actually, so I can at least have an idea about what's being done.

**Comment by vkarpov15** [comment]
See [here](https://github.com/vkarpov15/mgo/pull/1) for diff. The general idea is:

1) have a separate test suite, `RemoteSuite`, for tests that don't require a local mongod setup. With this, the Kerberos tests no longer have to use the `Setup()` and `TearDown`, so the tests can run on Windows.
2) I switched to using the official drivers Kerberos testing host instead of the mms one, because its the standard practice, and SSPI doesn't really do keytabs as far as I know.
3) Because SSPI requires you to acquire your kerberos ticket in-process, I added an environment variable for setting the password that gets passed to SSPI

**Comment by niemeyer** [comment]
Okay, that looks reasonable, but let's name it what it is: KerberosSuite, and have the suite defined right above kerberosFlag instead of having a new file for it. We can reorganize it after it lands if necessary.

If that's what it is, we can merge it in this PR so you can make sure tests pass before we merge this.

**Comment by vkarpov15** [comment]
Ok I merged in the test suite. Here's a screen cap of the windows tests passing:

image

**Comment by niemeyer** [comment]
Thanks! Can you please run the tests on Linux as well, to make sure we're still good there after the changes?

**Comment by vkarpov15** [comment]
Tests now pass on ubuntu 12.04 as well:

image

**Comment by vkarpov15** [comment]
Hey Gustavo, just a ping for any more comments

**Pull request merged**